### ISSUE TITLE
Latest 1.13.2 throws NoSuchFieldError: USE_FAST_DOUBLE_PARSER

### ISSUE THREAD

**Issue by lilgreenbird** [AzureAD/microsoft-authentication-library-for-java#560]
Our [JDBC driver](https://github.com/microsoft/mssql-jdbc) has recently upgraded to the latest v1.13.2 release as we require the latest updates to fasterxml.jackson.core:jackson-databind as component Azure Devops governance reported as a vulnerability.  However it looks like the latest has references to undefined USE_FAST_DOUBLE_PARSER.

The following is a test we have federated authentication which worked in prev releases:

```
        try {
            final PublicClientApplication clientApplication = PublicClientApplication.builder(fedauthClientId)
                    .executorService(Executors.newFixedThreadPool(1)).authority(stsurl).build();
            final CompletableFuture<IAuthenticationResult> future = clientApplication
                    .acquireToken(UserNamePasswordParameters.builder(Collections.singleton(spn + "/.default"),
                            azureUserName, azurePassword.toCharArray()).build());

            final IAuthenticationResult authenticationResult = future.get();

            secondsBeforeExpiration = TimeUnit.MILLISECONDS
                    .toSeconds(authenticationResult.expiresOnDate().getTime() - new Date().getTime());
            accessToken = authenticationResult.accessToken();
        } catch (Exception e) {
            fail(e.getMessage());
        }
```

This now results in:
```
org.opentest4j.AssertionFailedError: java.lang.NoSuchFieldError: USE_FAST_DOUBLE_PARSER
```

**Comment by Avery-Dunn** [comment]
Hello @lilgreenbird  : What version of our library were you using before, and what version of Java are you using for your test? I'm able to run your code on my machine without issue, and all of our own federated username/password tests are working.

Was anything else changed in the driver when you updated to v1.13.2?

**Comment by lilgreenbird** [comment]
we were on 1.13.0 but that got flagged by component governance as having a vulnerability. I tried 1.13.1 but that wasn't fixed so we went directly to 1.13.2.  This is only happening when we are running against a federation server the driver code hasn't changed. Here is the stack trace:

```
java.util.concurrent.ExecutionException: java.lang.NoSuchFieldError: USE_FAST_DOUBLE_PARSER
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at com.microsoft.sqlserver.jdbc.fedauth.FedauthCommon.getFedauthInfo(FedauthCommon.java:170)
	at com.microsoft.sqlserver.jdbc.fedauth.FedauthCommon.setupEachTest(FedauthCommon.java:120)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptLifecycleMethod(TimeoutExtension.java:126)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptBeforeEachMethod(TimeoutExtension.java:76)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeMethodInExtensionContext(ClassBasedTestDescriptor.java:506)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$synthesizeBeforeEachMethodAdapter$21(ClassBasedTestDescriptor.java:491)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachMethods$3(TestMethodTestDescriptor.java:171)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeMethodsOrCallbacksUntilExceptionOccurs$6(TestMethodTestDescriptor.java:199)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeMethodsOrCallbacksUntilExceptionOccurs(TestMethodTestDescriptor.java:199)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeEachMethods(TestMethodTestDescriptor.java:168)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
	at org.junit.platform.runner.JUnitPlatform.run(JUnitPlatform.java:151)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.lang.NoSuchFieldError: USE_FAST_DOUBLE_PARSER
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseFloatPrimitive(StdDeserializer.java:1073)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseFloatPrimitive(StdDeserializer.java:1049)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$FloatDeserializer.deserialize(NumberDeserializers.java:599)
	at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$FloatDeserializer.deserialize(NumberDeserializers.java:579)
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:177)
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4697)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3652)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3620)
	at com.microsoft.aad.msal4j.JsonHelper.convertJsonToObject(JsonHelper.java:33)
	at com.microsoft.aad.msal4j.UserDiscoveryRequest.execute(UserDiscoveryRequest.java:36)
	at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.processPasswordGrant(AcquireTokenByAuthorizationGrantSupplier.java:91)
	at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.execute(AcquireTokenByAuthorizationGrantSupplier.java:43)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:69)
	at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:18)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
```
We're on java 19 now but that is not relevant this is happening on java 17 and 11 as well. This works fine if I revert back to 1.13.0.

**Comment by pengfan8080** [comment]
Is this problem solved? I also encounter the same bug as the author. I don't know how to solve it.

**Comment by Avery-Dunn** [comment]
@lilgreenbird : I was using Java 8 before but just tried it with Java 17 and still didn't run into that issue. When you say you "tried 1.13.1 but that wasn't fixed", does that mean you ran into this "USE_FAST_DOUBLE_PARSER" issue in 1.13.1 also?

Could you also check to make sure there isn't some cached dependency weirdness going on? I just checked [the javadocs for 2.14.0-rc1](https://javadoc.io/doc/com.fasterxml.jackson.core/jackson-core/2.14.0-rc1/index.html), and the USE_FAST_DOUBLE_PARSER field is in the StreamReadFeature enum so I don't know why you'd be getting a NoSuchFieldError. However, that field only was added in version 2.14 and our library was using 2.13.2.1 before, so if an older version of jackson.core is cached somewhere then you'd definitely get that error.

Hello @pengfan8080 : Are you getting the same bug when using MSAL Java version 1.13.2? What version were you using before, and could you also check to make sure the 2.14.0-rc1 version of jackson.core is being used?

**Comment by lilgreenbird** [comment]
2 different things here - 1.13.1 did not fix the component governance issue that was in 1.13.0 so we upgraded to 1.13.2 which did fix the component issue but now has this USE_FAST_DOUBLE_PARSER problem.

I have updated the project the pom file and verified we are using the latest jackson-databind 2.14.0-rc1 from msal4j 1.13.2. Previous 1.13.0 version used jackson-databind 2.13.2.1 which did not have this problem but was flagged for vulnerability. 

Did you run the test against a federated server as mentioned? This doesn't happen when running against a regular on prem SQL Server. Please see stack trace included above on the code path the error originates from within MSAL code:

```
at com.microsoft.aad.msal4j.JsonHelper.convertJsonToObject(JsonHelper.java:33)
```

If we revert back to the old 1.13.0 version this goes away but as mentioned, that version has a vulnerability issue it gets flagged by component governance in Azure Devops I've filed a vulnerability report on this [VULN-078695](https://msrc.microsoft.com/report/vulnerability/VULN-078695).

Given this current issue, federated authentication in the driver is broken we will have to revert back to 1.13.0 for the time being until this is resolved.

**Comment by Avery-Dunn** [comment]
To be clear, by 'federated server' you just mean a server hooked up to some sort of federated active directory set up, right? Like ADFS or something similar? If so, that shouldn't make a difference since that convertJsonToObject is used throughout the library and not just for federated users, and that USE_FAST_DOUBLE_PARSER field isn't something we reference directly so I can't think of what our library could be doing to access it only in some scenarios.

The only way I can reproduce that NoSuchFieldError in MSAL Java is by explicitly putting conflicting versions of jackson-databind and jackson-core in my pom.xml. I'm not sure how JDBC is built and packaged, but is there anything different between the dependencies in a federated vs non-federated server?

I see that JDBC has azure-security-keyvault-secrets as a dependency, which has azure-core as a dependency which also has all of this jackson stuff but they're still on 2.13. We also have that as a dependency except it's just a 'test' dependency for us, so maybe Maven is still using 2.13 under the hood? Just to make sure the dependencies are all being used as expected, you can try adding the following to your pom.xml, to explicitly force Maven to only use version 2.14.0 of that jackson stuff in your project:
```
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>[2.14.0-rc1]</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-core</artifactId>
            <version>[2.14.0-rc1]</version>
        </dependency>
```
(I can only reproduce your error locally by setting jackson-core to anything less than 2.14.0-rc1)

**Comment by lilgreenbird** [comment]
hi @Avery-Dunn 

you are correct I see where the problem is now. The issue is that MSAL only has a dependency on jackson-databind but not jackson-core so when we upgraded to msal 1.13.2 jackson-databind gets bumpbed up to 2.14.0-rc1 since it overrides the older version included by azure-security-keyvault however not jackson-core so that still references the older version 
![image](https://user-images.githubusercontent.com/8549794/197054680-6091fd11-e909-42d5-a2d6-4bb6babf2458.png)

If I explicitly add a dependency to jackson-core and force the version to 2.14.0-rc1 then this will work. However, our driver does not depend on any of the jackson-* libraries and we don't want to add a new dependency just for this. If MSAL depends on both of these libraries and require them to be the same version it seems both of these should be included in your pom?

**Comment by Avery-Dunn** [comment]
This is the first time that we've had these sorts of issues where jackson-databind/jackson-core where out of sync enough to cause problems, and since we only use jackson-databind we've let jackson-core be a transient dependency. 

It seems like the root cause of the issue is that jackson-databind has kind of a soft requirement for version 2.14 of jackson-core, which means that Maven grabs jackson-core 2.13 to satisfy azure-core then we run into this issue. Since all of the jackson's 2.14 releases so far are just release candidates, I believe Maven will prefer the latest 2.13 release instead of any of the 2.14's.

Since neither of our projects relies on jackson-core the best option would be for azure-core to update their project to use one of the 2.14 release candidates. It looks like they've started creating PRs to update their use of jackson, but I'm not sure what they're release schedule is: https://github.com/Azure/azure-sdk-for-java/pull/31559

I'll try to figure out when they're planning a release with these dependencies updated. If it seems like it'll be soon then I think it'd be best to let that solve this issue, otherwise we'll add jackson-core as a dependency in MSAL Java and we'll probably get a hotfix out sometime next week.

**Comment by lilgreenbird** [comment]
thanks, in the meantime I've reverted the driver back to the old version of msal4j.

**Comment by Avery-Dunn** [comment]
Hello @lilgreenbird : In our recent [1.13.3](https://mvnrepository.com/artifact/com.microsoft.azure/msal4j/1.13.3) release we're now using version 2.13.4.2 of jackson-databind, which is a version that has the security fix and is compatible with other 2.13 versions of jackson-core

You should be able to upgrade to 1.13.3 of MSAL Java to solve this USE_FAST_DOUBLE_PARSER issue and the security vulnerability. Closing this issue, feel free to reopen/leave a comment if you're still running into issues.

**Comment by pengfan8080** [comment]
Thank you. I solved this problem by upgrading the version. Thank you very much
